go-controller/pkg/ovn/ovn.go

```diff
@@ -59,20 +57,57 @@ type ACLLoggingLevels struct {
 	Deny  string `json:"deny,omitempty"`
 }
 
-// Controller structure is the object which holds the controls for starting
-// and reacting upon the watched resources (e.g. pods, endpoints)
-type Controller struct {
+// controllerInfo structure is place holder for all fields shared among controllers.
+type controllerInfo struct {
 	client       clientset.Interface
 	kube         kube.Interface
 	watchFactory *factory.WatchFactory
-	stopChan     <-chan struct{}
+	podRecorder  *metrics.PodRecorder
+
+	// event recorder used to post events to k8s
+	recorder record.EventRecorder
+
+	// libovsdb northbound client interface
+	nbClient libovsdbclient.Client
+
+	// libovsdb southbound client interface
+	sbClient libovsdbclient.Client
+
+	// has SCTP support
+	SCTPSupport bool
+}
+
+type Controller interface {
+	Start(ctx context.Context) error
+	Stop()
+
+	GetInternalCacheEntry(eventObjType reflect.Type, obj interface{}) interface{}
+
+	AddResource(eventObjType reflect.Type, obj interface{}, fromRetryLoop bool, extraParameters interface{}) error
+	UpdateResource(eventObjType reflect.Type, oldObj, newObj interface{}, inRetryCache bool, extraParameters interface{}) error
+	DeleteResource(eventObjType reflect.Type, obj, cachedObj, extraParameters interface{}) error
+	GetSyncFunc(reflect.Type) (func([]interface{}) error, error)
+
+	RecordAddEvent(eventObjType reflect.Type, obj interface{})
+	RecordUpdateEvent(eventObjType reflect.Type, obj interface{})
+	RecordDeleteEvent(eventObjType reflect.Type, obj interface{})
+	RecordSuccessEvent(eventObjType reflect.Type, obj interface{})
+	RecordErrorEvent(eventObjType reflect.Type, err error, reason string, obj interface{})
+}
+
+// DefaultL3Controller structure is the object which holds the controls for starting
+// and reacting upon the watched resources (e.g. pods, endpoints) for default l3 network
+type DefaultL3Controller struct {
```

Comment:
I think that the type definition and some basic methods like `NewDefaultL3Controller` and `initRetryFrameworkForMaster` can go now to default_l3_controller.go as they are likely to stay specific to this controller type.

Moving forward I think other methods scoped in DefaultL3Controller will either be move to free utility methods or to a base type shared by all network controllers types; and the specific network controller types will have their own package.